ai_generation_functions.py

```diff
@@ -10,19 +12,49 @@
 from requests.exceptions import Timeout, RequestException
 import gradio as gr
 from prompts import TITLE_GENERATION_PROMPT, LYRICS_GENERATION_PROMPT, SONG_STYLE_GENERATION_PROMPT, LYRICS_ANALYSIS_SONG_STYLE_PROMPT, LYRICS_ANALYSIS_INSTRUMENTS_PROMPT, music_categories
-
+import tempfile
+import PyPDF2
 
 load_dotenv()
 client = OpenAI()
 
 @observe()
 def get_lyrics(youtube_link):
-    try:
-        with open('tests/ly_1.txt', 'r', encoding='utf-8') as f:
-            lyrics = f.read()
-        return lyrics if lyrics else "No lyrics found."
-    except Exception as e:
-        return f"An error occurred: {str(e)}"
+        video_id = extract.video_id(youtube_link)
+        # Try to get the transcript (subtitles) in English
+        try:
+            transcript_list = YouTubeTranscriptApi.list_transcripts(video_id)
+            transcript = transcript_list.find_transcript(['en'])
+            transcript_data = transcript.fetch()
+            lyrics = ' '.join([entry['text'] for entry in transcript_data])
+            return lyrics if lyrics else "No lyrics found."
+        except Exception as e:
+            print(f"No subtitles found or error in fetching subtitles: {str(e)}")
+            print("Proceeding to download audio and transcribe using Whisper.")
+
+            # Download YouTube audio
+            yt = YouTube(youtube_link, on_progress_callback=on_progress)
+            audio = yt.streams.get_audio_only()
+            audio.download(filename="temp", mp3=True)
+
+            # Transcribe audio using Whisper
+            model = whisper.load_model("base")
+            result = model.transcribe("temp.mp3")
+
+            # Save transcription result to file
+            with open("lyrics.txt", "w", encoding="utf-8") as f:
+                f.write(result["text"])
+
+            # Read and return lyrics
+            with open("lyrics.txt", "r", encoding="utf-8") as f:
+                lyrics = f.read()
```

Comment:
The result from Whisper is the lyrics, so why not use it directly instead of reading the temp file again?